libflux/flux-core/src/semantic/version.rs

```diff
@@ -0,0 +1,139 @@
+//! Utility to compare Flux types with their version.
+
+use std::cmp::max;
+
+use crate::semantic::types::{MonoType, PolyType};
+
+/// VersionDiff represents a difference between two types.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub enum VersionDiff {
+    /// Represents a backwards incomptible change to the types
+    Major = 2,
+    /// Represents a backwards comptible change to type
+    Minor = 1,
+    /// Represents no change to the types
+    Patch = 0,
+}
+
+/// Report the version difference between the old and new polytypes.
+pub fn diff(old: &PolyType, new: &PolyType) -> VersionDiff {
+    if old == new {
+        return VersionDiff::Patch;
+    }
+
+    // TODO handle constraints
+
+    diff_monotype(&old.expr, &new.expr)
+}
+
+fn diff_monotype(old: &MonoType, new: &MonoType) -> VersionDiff {
+    if old == new {
```

Comment:
Nit: This isn't really a performance sensitive code, but doing `==` at every level of the `Monotype` makes it `O(N^2)`